### DISCUSSION TITLE
Mutable type in default of index_property shares between objects.

### DISCUSSION THREAD

**Discussion by artofhuman** [sqlalchemy/sqlalchemy#12137] [Ideas]
[index_property](https://docs.sqlalchemy.org/en/20/orm/extensions/indexable.html#sqlalchemy.ext.indexable.index_property)

Does not allow to pass mutable types as default safety. Because this object shares between instance.

```
class User(Base):
   data: Mapped[dict] = mapped_column(JSONB)

   ident = index_property("data", "ident", default={})


u1 = User()
u2 = User()

u1.ident["foo"] = "bar"


print(u1.ident)
# => {"foo": "bar"}

print(u2.ident)
# => {"foo": "bar"}


```

I know it may be fixed in project by subclassing of index_property, but would be nice to have ability to pass callable objects or have api similar to dataslasses - `field(default_factory=dict)` in lib

**Comment by zzzeek** [comment]
I have not really ever looked at indexable but it appears to be stateless so the default value can't be mutable in any case, as it's discarded immediately

```py
class User(Base):
   data: Mapped[dict] = mapped_column(JSONB)

   ident = index_property("data", "ident", default_factory=dict)

u1 = User()
u1.ident['foo'] = 'bar'

assert 'foo' in u1.ident   # <--- would fail, this is a new dictionary

```
